### PR TITLE
supports `mimalloc` for ARC/ORC (1.15x ~ 1.22x speedup booting the compiler)

### PR DESCRIPTION
**Pull request by ringabout**
Credits to @Yardanico
ref https://github.com/Yardanico/mimalloc_nim

I added the commit by
git commit --amend --author="Danil Yarantsev <tiberiumk12@gmail.com>" -m "supports mimalloc for ARC/ORC"

### CODE REVIEW

**Comment by ringabout** [comment]
The first step is to check whether it reduces booting compilation time. I will refactor it later. Making it work comes first.

**Comment by ringabout** [comment]
It is weird that importing `std/os` in the global Nim config increases 200 MiB memory usage.

**Comment by ringabout** [comment]
The booting time seems to drop at some extent by testing locally and on the CI

[ubuntu-20.04 (batch: 0_3)](https://github.com/nim-lang/Nim/actions/runs/3057985702/jobs/4933698502#logs) 15.347s => 12.546s （1.22x)
[ubuntu-20.04 (batch: 1_3)](https://github.com/nim-lang/Nim/actions/runs/3057985702/jobs/4933698538#logs) 9.140s => 7.867s  (1.16x)
[ubuntu-20.04 (batch: 2_3)](https://github.com/nim-lang/Nim/actions/runs/3057985702/jobs/4933698570#logs) 12.374s => 10.688s (1.15x)

**Review by ghost — COMMENTED** [review]
I think it's still beneficial to support dynamic mimalloc, because if you have multiple Nim DLLs, or the main Nim program and other Nim modules as shared libraries, if each of them has their own copy of mimalloc, you won't be able to share GC'd objects between the instances. So, similar to how nimrtl has it, we can just allow users to use dynamic mimalloc

**Comment by planetis-m** [comment]
There might be a problem with diagnostic tools like [heaptrack](https://github.com/KDE/heaptrack) and sanitizers that work best with `-d:useMalloc`. As I understand this switch is now abused to mean mimalloc. Does this PR still allow them to work reliably? Have you tested? ie try to leak memory and see if LeakSanitizers errors.

**Comment by ringabout** [comment]
> There might be a problem with diagnostic tools like [heaptrack](https://github.com/KDE/heaptrack) and sanitizers that work best with -d:useMalloc. As I understand this switch is now abused to mean mimalloc. Does this PR still allow them to work reliably? Have you tested?

Not really, I add a new `--d:useMimAlloc` switch. `-d:useMalloc` still works.

File: koch.nim [hunk 1]
```diff
@@ -606,6 +606,8 @@ proc runCI(cmd: string) =
     # the BSDs are overwhelmed already, so only run this test on the other machines:
     kochExecFold("Boot Nim ORC", "boot -d:release --mm:orc --lib:lib")
 
+    kochExecFold("Boot Nim ORC with mimalloc", "boot -d:release --mm:orc --lib:lib -d:useMalloc")
```
Comment (ringabout) [inline koch.nim]:
```suggestion
    kochExecFold("Boot Nim ORC with mimalloc", "boot -d:release --mm:orc --lib:lib -d:useMimalloc")
```

**Comment by ringabout** [comment]
Waiting for https://github.com/nim-lang/Nim/pull/19972

**Comment by Araq** [comment]
This is critical for the v2.0 release btw.

**Comment by ringabout** [comment]
Sure.

**Comment by Araq** [comment]
Closing for the time being as our allocator is also now threadsafe.

**Pull request closed without merging**